### PR TITLE
Feature/issue 55 config validation new config

### LINKED ISSUE #55: Implement Configuration Validation Step
#37 

Issue Description
---
Some errors within configuration, such as the date range for swap positions and the cashflow arguments for cashflows are only checked towards the end of generation. Should there be an issue with the user-set values in these sections an error will only be raised once other generation has completed. Time spent generation up to the point an error is raised would be redundant as any subsequent generation after fixing said issues would supersede the already-generated data. 

Design
---
Implement within the primary app.py, before any generation logic, a verification step of the configuration file where these edge cases are assessed before any operation begins. This step should return a list of all errors in the configuration rather than just the first one found.

Things to check are:
1. All objects have a record count, maximum number of objects per file, and valid file type.
2. Record count is hreater than 0
3. Swap Contracts have a valid range for swaps per counterparty.
4. Swap Positions have valid start/end date values.
5. Swap Positions have a valid range for instruments per swap.
6. Cashflows have at least one definition in "cashflow_generation"
7. All defined cashflows are valid (cashFlowType, cashFlowAccrual, cashFlowAccrualProbability and cashFlowPaydatePeriod are valid)

Documentation Changes
---
Documentation of the implemented method(s) adhering to the Numpy standard as per the rest of the project.

Test Evidence
---
No further regression testing needs implementing.
Tests verifying that each of the above cases are caught and reported. Provide example config.json's across all of which all above issues are caught and reported to the user.

Validation in Develop
---

### PR DESCRIPTION
**Pull request by GalaJoe**
This is the PR for changes made in the comments of #55. It has the checks for swap data removed from it whilst we restructure the project. 

This closes #55

### CODE REVIEW

File: src/app.py [hunk 1]
```diff
@@ -33,10 +33,12 @@
 import importlib
 import ujson
 import os
+import sys
 from argparse import ArgumentParser
 from utils.sqlite_database import Sqlite_Database
 from multi_processing.coordinator import Coordinator
 import multi_processing.batch_size_calc as batch_size_calc
+import utils.config_validator as config_validator
```
Comment (tonasanya) [inline src/app.py]: create a package named validator and put the config validator here

Comment (GalaJoe) [reply src/app.py]: Done

File: src/utils/config_validator.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            raise ConfigError()
+
+    except ConfigError:
+        print("Issue(s) within Config:")
+        for error in errors:
+            print(error)
+        raise
+
+    else:
+        print("No issues within Config")
```
Comment (tonasanya) [inline src/utils/config_validator.py]: Validate should return a ValidationResult.
ValidationResult should have a boolean success and an Array Errors[].
You can then test the result and get the error messages out.

The error should be thrown from App.py based on the result of the validation

File: tests/test_validation/config_validation_test.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def run_test(test_name):
+    success_config = load_file(test_name+"_success.json")
+    failure_config = load_file(test_name+"_fail.json")
+    success_result = run_validation(success_config).check_success()
+    fail_result = run_validation(failure_config).check_success()
+    assert success_result and not fail_result
+
+def load_file(file_name):
```
[the remark this answers is not in the document]
Comment (GalaJoe) [reply tests/test_validation/config_validation_test.py]: there is 2

**Comment by GalaJoe** [comment]
Remaining hound issues are irrelevant/unavoidable as per our prior testing approach.

**Comment by GalaJoe** [comment]
Merge conflicts resolved. Remaining hound issues are akin to those ignored in prior testing implementation.

File: tests/test_validation/config_validation_test.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import sys
+sys.path.insert(0, 'src/')
+from validator import config_validator as validator 
+import json
+
+
+def test_record_count():
```
Comment (tonasanya) [inline tests/test_validation/config_validation_test.py]: -put comments on each test to describe what it does
-each test should process one file and be named after the file/ expected result e.g. test_pool_size_too_small_fails_validation: This test should have the appropriate file name declared fully as a string in it. 
-I've changed my mind, each test should be fully self contained and assert success = true or assert = false as appropriate. There should be no run_tests or run_validation methods. The load_file can stay though as that's a useful abstraction

Comment (GalaJoe) [reply tests/test_validation/config_validation_test.py]: done

**Review by tonasanya — APPROVED** [review]

**Pull request merged**